# load libraries and data
```{r}
library(dplyr)
library(readxl)
'%notin%' <- Negate('%in%')

# read in MOTU table by sample (Swarm recount) 
mt <- read.csv("data/merged.uni.c10.140.190.sht.vsc.srt.chi.sin.sw1_output.counts.csv") %>%
  select(id, starts_with("sample"))

# read in list of NJ vertebrates
njv <- read.csv("data/nj_vertebrates_taxonomy.csv") %>%
  left_join(select(read.csv("data/nj_vertebrates.csv"), 
                   mol_name = scientific_name, common_name)) %>%
  bind_rows(data.frame(mol_name = "none", ncbi_name = "none", 
                       common_name = "none", class = "none"))

# read in total reads & taxonomy per MOTU (ecotag output)
f <- read.table("data/FishIBI.2023.final.MOTU.table.MiFish.tsv", 
                sep="\t", header=TRUE) %>%
  select(-definition) %>%
  rename(bestid = starts_with("best_id")) %>%
  left_join(mt, by = "id") %>%
  select(seqid = id, bestid, order = order_name, family = family_name, genus = genus_name, species = species_name, count = size, sci = scientific_name, splist = starts_with("species_list"),
         starts_with("sample"), sequence); rm(mt)

# read in sample names and metabarcoding run information
s <- read_xlsx("data/2024 Metabarcoding NJDEP Batch 8 022924.xlsx",
               sheet = "Sheet1") %>%
  mutate(sample = paste0("sample.",`Seq. Run X, Y, Z`, `Index Set`, 
                       substr(`Sample Index Code`,1,1), "0",
                       substr(`Sample Index Code`,2,2)))

# make a lookup table between run/set/well sample number and real sample names
samp_lookup <- select(s, sample, t = Sample_ID, 
                      Sample_Type, Date = `Date Collected`,
                      Stream) %>%
  mutate(real_names = gsub(pattern = " ", replacement = "_", x = t))

# read in BLASTed MOTUs
# more info on fields: https://www.metagenomics.wiki/tools/blast/blastn-output-format-6
options(digits = 4)
b <- read.csv("data/all.MOTUs.blasted.txt", header = F) %>%
  rename(qseqid = 1, sseqid = 2, pident = 3, length = 4, mismatch = 5, gapopen = 6,
         qstart = 7, qend = 8, sstart = 9, send = 10, evalue = 11, bitscore = 12,
         qlen = 13, slen = 14, staxids = 15, sscinames = 16, scomnames = 17, 
         sskingdoms = 18) %>%
  mutate(qcov = length/qlen,
         pident = pident/100) %>%
  filter(qcov >= 0.95 &
         pident >= 0.95) %>%
  left_join(njv, by = c("sscinames" = "ncbi_name"))

# summarize blast results
options(digits = 4)
bsum <- b %>%
  group_by(qseqid, sscinames, common_name) %>%
  summarize(bestmatch = round(max(pident), 3)) %>%
  ungroup() %>%
  arrange(qseqid, desc(bestmatch)) %>%
  group_by(qseqid) %>%
    mutate(names = paste(unique(sscinames), collapse = "/"),
           comnames = paste(unique(common_name), collapse = "/"),
           full.blast.bestmatch.each = paste(bestmatch, collapse = "/"),
           full.blast.bestmatch = max(bestmatch)) %>%
  ungroup() %>%
  select(seqid = qseqid, blast.names = names, 
         full.blast.comnames = comnames, full.blast.bestmatch, full.blast.bestmatch.each) %>%
  distinct()

# subset blast results to local species only
blocal <- b %>%
  filter(sscinames %in% njv$ncbi_name)

# summarize blast results
options(digits = 4)
bsum.local <- blocal %>%
  group_by(qseqid, sscinames, common_name) %>%
  summarize(bestmatch = round(max(pident), 3)) %>%
  ungroup() %>%
  arrange(qseqid, desc(bestmatch)) %>%
  group_by(qseqid) %>%
    mutate(names = paste(unique(sscinames), collapse = "/"),
           comnames = paste(unique(common_name), collapse = "/"),
           bestmatch.each = paste(bestmatch, collapse = "/")) %>%
  ungroup() %>%
  select(seqid = qseqid, local.blast.names = names, 
         local.blast.comnames = comnames, local.blast.bestmatch.each = bestmatch.each) %>%
  distinct()

# assign real names to sample columns
cols <- data.frame(sample = colnames(select(f, starts_with("sample.")))) %>%
  left_join(samp_lookup, by = "sample") %>%
  mutate(real_names = ifelse(is.na(real_names), sample, real_names))

# samples not in sequencing files

test = data.frame(sample = 
                     samp_lookup$sample[samp_lookup$sample %notin% colnames(f)]) %>%
  left_join(samp_lookup)
write.csv(test, "output/samples_not_in_sequencing_files.csv", row.names = F)
rm(test)

# change sample codes in raw reads table to real sample names
colnames(f) # check to make sure the right columns are being changed!
colnames(f)[10:156] <- cols$real_names

# select out just the negatives
fn_step1 <- f %>%
  select(seqid, order, family, genus, species, bestid, sci, splist,
         count, starts_with("ENC"), starts_with("Forecp"), starts_with("PCR"), ends_with("-B"),
         ends_with("RR"), sequence)
colnames(fn_step1)
fn_step1$total_in_neg <- apply(fn_step1[,10:38],1, sum)
fn_step1$max_in_neg <- apply(fn_step1[,10:38],1, max)
fn <- fn_step1 %>%
  tidyr::replace_na(list(species="none")) %>%
  left_join(select(njv, species = ncbi_name, ecotag.common = common_name), 
            by = "species") %>%
  left_join(bsum.local, by = "seqid") %>%
  left_join(bsum, by = "seqid") %>%
  mutate(final.common = case_when(species=="none" ~ local.blast.comnames,
                                 TRUE ~ ecotag.common),
         final.names = case_when(species=="none" ~ local.blast.names,
                                 TRUE ~ species),
         final.match = case_when(species=="none" ~ local.blast.bestmatch.each,
                                 TRUE ~ as.character(bestid))) %>%
  select(seqid, order, family, genus, species, final.names, final.match,
         ecotag.match = bestid, 
         full.blast.bestmatch,
         local.blast.bestmatch.each, 
         full.blast.bestmatch.each,
         ecotag.name = sci, local.blast.names, full.blast.names = blast.names,
         count, total_in_neg, max_in_neg, 
         starts_with("ENC"), starts_with("Forecp"), 
         starts_with("PCR"), ends_with("-B"),
         ends_with("RR"), sequence); rm(fn_step1)

# sum the reads across all negatives
sum(select(fn, starts_with("ENC"), starts_with("Forecp"), starts_with("PCR"), ends_with("-B"),
         ends_with("RR"))) # 209990 (188343 last year)

# select out just the samples
fs_step1 <- f %>%
  select(seqid, order, family, genus, species, bestid, sci, splist,
         count, starts_with("eDNA"), sequence)  %>%
  select(-ends_with("-B"), -ends_with("RR"))
colnames(fs_step1)
fs_step1$total_in_samps <- apply(fs_step1[,10:85],1, sum)
fs_step1$total_in_neg <- fn$total_in_neg
fs_step1$max_in_neg <- fn$max_in_neg
fs <- fs_step1 %>%
  tidyr::replace_na(list(species="none")) %>%
  left_join(select(njv, species = ncbi_name, ecotag.common = common_name), 
            by = "species") %>%
  left_join(bsum.local, by = "seqid") %>%
  left_join(bsum, by = "seqid") %>%
  mutate(final.common = case_when(species=="none" ~ local.blast.comnames,
                                 TRUE ~ ecotag.common),
         final.names = case_when(species=="none" ~ local.blast.names,
                                 TRUE ~ species),
         final.match = case_when(species=="none" ~ local.blast.bestmatch.each,
                                 TRUE ~ as.character(bestid))) %>%
  select(seqid, order, family, genus, species, final.common, final.names, final.match,
         ecotag.match = bestid, 
         full.blast.bestmatch,
         local.blast.bestmatch.each, 
         full.blast.bestmatch.each,
         ecotag.name = sci, local.blast.names, full.blast.names = blast.names,
         count, total_in_samps, total_in_neg, max_in_neg, 
         starts_with("eDNA"), sequence); rm(fs_step1)
  
# sum the reads across all samples
sum(select(fs, starts_with("eDNA"))) # 5089075 (8005382 last year)

f_to_curate <- fs %>%
  left_join(select(fn, seqid, starts_with("ENC"), starts_with("Forecp"), 
                   starts_with("PCR"), ends_with("-B"),
                   ends_with("RR")), by = "seqid") %>%
  mutate(finalID_source = case_when(species=="none" ~ "BLAST (NCBI eukaryotes)",
                                    TRUE ~ "Reference database (ecotag)")) %>%
  select(1:5, finalID_source, 6:126) %>%
  mutate(class = case_when(order %in% c("Anguilliformes", "Centrarchiformes",
                                          "Cypriniformes", "Cyprinodontiformes",
                                        "Esociformes", "Perciformes",
                                        "Salmoniformes", "Siluriformes") ~ "Actinopteri",
                           order %in% c("Anseriformes", "Passeriformes") ~ "Aves",
                           order %in% c("Artiodactyla", "Carnivora", 
                                        "Chiroptera", "Didelphimorphia",
                                        "Eulipotyphla", "Lagomorpha",
                                        "Primates", "Rodentia") ~ "Mammalia",
                           order %in% c("Caudata") ~ "Amphibia",
                           order %in% c("Testudines") ~ "Testudines")) %>%
  arrange(class, final.names, desc(total_in_samps))

# write.csv(f_to_curate, file = "output/Step1_raw_fishIBI2023_MOTUs_to_curate.csv",
#           row.names = F)

```
# Step 2: manually curate taxonomy assignments
Save the file you just exported called Step1_raw_fishIBI2023_MOTUs_to_curate.csv and rename it to Step2_raw_fishIBI2023_MOTUs_curated.csv. If opening in Excel, first open a blank sheet and import the csv file as text that ensures the blast.each columns are imported as characters, otherwise cells with "1/1" will be converted to "1-Jan". Go through each finalID and check to make sure that blast hasn't returned a much better match (e.g., 2 % points, 100% vs. 98%). The important ones are those where ecotag had a poor match (say 85 or 90%) whereas Blast had a "good" match (say 98 or 100%). Those are clear cases of a sequence that was inadvertently omitted from the reference dataset (filtered out during download or in one of the cleaning & taxonomy associating steps). Go through each finalID and update it where needed to the BLAST identity. Be sure to update the finalID_source, final.match, and the full higher taxonomy too. (Note: the higher taxonomy could also be updated programmatically for those, but I've been doing it manually as updated higher taxonomy has been rare.) Save the curated excel file as a csv named: "Step2_raw_fishIBI2023_MOTUs_curated.csv".
# Create "final_raw_data.csv" from curated data file
After the finalID and associated column is finalized, format the data into its final form.
```{r}
colnames(f_to_curate) # to modify the column selection part below; 
f2 <- read.csv("output/Step2_raw_fishIBI2023_MOTUs_curated.csv") %>%
  select(seqid, finalID = final.names, finalID_names = final.common,
         bestmatch = final.match, finalID_source, finalID_class = class,
         finalID_order = order, finalID_family = family, ecotagmatch = ecotag.match, 
         bestblastmatch = full.blast.bestmatch, ecotagID = ecotag.name,
         full.blast.names, full.blast.bestmatch.each) %>%
  left_join(select(f_to_curate, seqid, total_in_samps:`eDNA072-071323-B_RR`), by = "seqid") %>%
  select(seqid, total_in_samps, total_in_neg, max_in_neg,
         finalID, finalID_names, bestmatch, finalID_source, finalID_class, 
         finalID_order, finalID_family, starts_with("eDNA"), 
         sequence, ecotagmatch, 
         bestblastmatch, ecotagID,
         full.blast.names, full.blast.bestmatch.each)  %>%
  select(-ends_with("-B"), -ends_with("RR")) %>%
  left_join(select(fn, seqid, starts_with("ENC"), starts_with("Forecp"), 
                   starts_with("PCR"), ends_with("-B"),
                   ends_with("RR")), by = "seqid") %>%
  mutate(finalID = case_when(bestmatch < 0.95 ~ "(no reliable matches)",
                             is.na(finalID) ~ "(no reliable matches)",
                             TRUE ~ finalID)) %>%
  mutate(finalID_names = case_when(finalID == "(no reliable matches)" ~ 
                                     "(no reliable matches)",
                             TRUE ~ finalID_names)) %>%
  mutate(sort = case_when(finalID == "(no reliable matches)" ~ "z",
                          TRUE ~ finalID_family)) %>%
  arrange(finalID_class, finalID_order, sort, finalID, desc(total_in_samps)) %>%
  select(-sort)

# write.csv(f2, "output/2023_final_raw_data.csv", row.names = F)
```
# Create "final_sample_table.csv"
```{r}
f3 <- f2 %>%
  select(seqid, finalID, finalID_names, finalID_class, finalID_order, 
         finalID_family, bestmatch, total_in_samps, total_in_neg, starts_with("eDNA")) %>%
  select(-ends_with("-B"), -ends_with("RR")) %>%
  left_join(select(fn, seqid, starts_with("ENC"), starts_with("Forecp"), 
                   starts_with("PCR"), ends_with("-B"),
                   ends_with("RR")), by = "seqid") %>%
  mutate(sort = case_when(finalID == "(no reliable matches)" ~ "z",
                          TRUE ~ finalID_family)) %>%
  group_by(sort, finalID, finalID_names, finalID_class, 
           finalID_order, finalID_family) %>%
  summarize(bestmatch = max(bestmatch),
            total_in_samps = sum(total_in_samps),
            total_in_neg = sum(total_in_neg),
            across(#total_in_samps, total_in_neg, 
                   'eDNA097a-060523-03':'eDNA072-071323-B_RR', sum),
            .groups = "drop") %>%
  arrange(finalID_class, finalID_order, sort, finalID, desc(total_in_samps)) %>%
  select(-sort)

# write.csv(f3, "output/2023_final_sample_table.csv", row.names = F)
  
```

# make species lists by sample
```{r}
final <- f3 %>%
  tidyr::pivot_longer(cols = `eDNA097a-060523-03`:`eDNA072-071323-B_RR`,
                      names_to = "ActivityID",
                      values_to = "Reads") %>%
  left_join(select(samp_lookup, ActivityID = real_names, 
                   SampleType = Sample_Type, Date, Stream), by = "ActivityID") %>%
  select(SampleType, Stream, ActivityID, Date, Class = finalID_class,
         SpeciesLatin = finalID, CommonName = finalID_names, 
         BestMatch = bestmatch, Reads) %>%
  filter(Reads>0) %>%
  mutate(BestMatch = case_when(SpeciesLatin == "(no reliable matches)" ~ NA,
                               TRUE ~ BestMatch),
         SpeciesLatin = case_when(SpeciesLatin == "(no reliable matches)" ~ 
                                    "No reliable matches",
                               TRUE ~ SpeciesLatin),
         CommonName = case_when(CommonName == "(no reliable matches)" ~ 
                                  "No reliable matches",
                               TRUE ~ CommonName)) %>%
  arrange(SampleType, ActivityID, Class, desc(Reads))

# write.csv(final, "output/2023_final_summary_table.csv", row.names = F)

```
